src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,227 @@
+package duke;
+
+import duke.common.Const;
+import duke.task.Task;
+
+import java.util.regex.Pattern;
+import java.time.format.DateTimeParseException;
+import java.time.LocalDateTime;
+
+/**
+ * Parser class handles and process user's inputs with commands
+ */
+public class Parser {
+
+    private TaskList taskList;
+    private Ui ui;
+    private boolean isExit;
+
+    /**
+     * Create an instance of Parser
+     *
+     * @param taskList taskList to get information of tasks
+     * @param ui       to interact with the users after processing the input
+     */
+    public Parser(TaskList taskList, Ui ui) {
+        this.taskList = taskList;
+        this.ui = ui;
+        this.isExit = false;
+    }
+
+    public boolean commandIsExit() {
+        return isExit;
+    }
+
+    /**
+     * Main logic in handling user input based on commands entered.
+     *
+     * @return response after processing the input.
+     */
+    public String parseInput(String input) throws DukeException {
+        try {
+            String[] commandLine = parseCommandLine(input);
+            Command command = Command.getCommand(commandLine[0]);
+            if (command == null) {
+                throw new DukeException.DukeInvalidCommandException();
+            }
+
+            String output;
+            switch (command) {
+            case BYE:
+                parseBye(commandLine);
+                output = ui.endMessage();
+                break;
+            case LIST:
+                parseList(commandLine);
+                output = ui.listMessage();
+                break;
+            case MARK:
+                parseMark(commandLine);
+                output = ui.markMessage(Integer.parseInt(commandLine[1]));
+                break;
+            case UNMARK:
+                parseUnMark(commandLine);
+                output = ui.unMarkMessage(Integer.parseInt(commandLine[1]));
+                break;
+            case ADD:
+                parseAdd(commandLine);
+                output = ui.addTaskMessage();
+                break;
+            case DELETE:
+                String toDelete = parseDelete(commandLine);
+                output = ui.deleteMessage(toDelete);
+                break;
+            case FIND:
+                String found = parseFind(commandLine);
+                output = ui.findMessage(found);
+                break;
+            default:
+                output = "Opps! something went wrong";
```

Comment:
Perhaps you could have this in a Ui method to print for unexpected errors?

Reply:
Yes. Thanks for pointing it out.